### PR TITLE
Feature: Forex websocket client

### PR DESCRIPTION
**Pull request by jengjeng**

### CODE REVIEW

File: api/forex/client.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	return &Client{
+		ID:      id,
+		WS:      ws,
+		Streams: streamsMap,
+		logger:  logger,
+	}, nil
+}
+
+// Connect to Forex websocket endpoint
+func (c *Client) Connect(listener func(*PriceResponse)) error {
```
Comment (dnfd) [inline api/forex/client.go:66]: I do not see how to handle reconnect with this api, can you add example?

File: api/forex/client.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+		return err
+	}
+
+	delete(c.Streams, market)
+
+	return nil
+}
+
+func (c *Client) getPayload(action string, data interface{}) *[]interface{} {
+	return &[]interface{}{
```
Comment (dnfd) [inline api/forex/client.go:141]: Pointer to slice again

File: api/forex/client.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		data,
+	}
+}
+
+func (c *Client) parsePriceData(message interface{}) (*PriceResponse, error) {
+	if res, ok := message.([]interface{}); ok {
+		code := res[0].(float64)
+		rType := res[1].(string)
+		if code == responseCode && rType == priceResponseType { // TODO: need to handle other response types later.
+			data := res[2].([]interface{})
```
Comment (dnfd) [inline api/forex/client.go:154]: it panics if data is not string/float

File: api/forex/client.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			pRes := &PriceResponse{
+				Market:    data[0].(string),
+				Price:     data[1].(string),
+				CreatedAt: data[2].(float64),
+				UpdatedAt: data[3].(float64),
```
Comment (dnfd) [inline api/forex/client.go:159]: panics on wrong type

File: websocket/websocket.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+// Client for forex api instance
+type Client struct {
+	Address   *string
```
Comment (dnfd) [inline websocket/websocket.go:16]: why pointer to string

File: websocket/websocket.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	Conn      *websocket.Conn
+	Listeners *[]func(interface{})
```
Comment (dnfd) [inline websocket/websocket.go:18]: pointer to slice

File: websocket/websocket.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		Address:   address,
+		logger:    logger,
+		Listeners: &[]func(interface{}){},
+	}, nil
+}
+
+// Connect to websocket endpoint
+func (client *Client) Connect(query string) error {
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
```
Comment (dnfd) [inline websocket/websocket.go:40]: Remove

File: websocket/websocket.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+				for _, listener := range *client.Listeners {
+					go listener(message)
+				}
+			}()
+		}
+	}()
+
+	// states handler
+	go func() {
+		for {
```
Comment (dnfd) [inline websocket/websocket.go:82]: should not be a loop

**Review by dnfd — CHANGES_REQUESTED** [review]

File: websocket/websocket.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	}, nil
+}
+
+// Connect to websocket endpoint
+func (client *Client) Connect() error {
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	// connection
+	u := url.URL{
```
Comment (dnfd) [inline websocket/websocket.go:51]: Faced a problem here, I'm trying to mock forex service in test, I need to pass this url, `ws://localhost:4444/ws/forex`, the problem is I cannot set path and if I pass the whole address its encoded. Also I cannot set wss, too.
If you want to use url type, take this as an argument to `New`

File: api/forex/client.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	return &[]interface{}{
+		1,
+		c.ID,
+		action,
+		data,
+	}
+}
+
+func (c *Client) parsePriceData(message interface{}) (*PriceResponse, error) {
+	if res, ok := message.([]interface{}); ok {
```
Comment (dnfd) [inline api/forex/client.go:156]: prefer guard clauses for less  nesting

File: api/forex/client.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	c.WS.AddListener(func(message interface{}) {
+		data, err := c.parsePriceData(message)
+		if err != nil {
+			c.logger.Println(err)
+		} else if data != nil {
+			listener(data)
+		}
+	})
+
+	markets := &[]string{}
```
[the remark this answers is not in the document]
Comment (jengjeng) [reply api/forex/client.go]: My first thought was I used `markets` as a variable in the client struct then I used it as a pointer. I will check it back to the normal slice.

File: api/forex/client.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+	UpdatedAt float64
+}
+
+// New Forex client
+func New(id string, url string, streams []string) (*Client, error) {
+	if url == "" {
+		return nil, errors.New("'url' is missing")
+	}
+	if streams == nil {
+		streams = []string{}
```
[the remark this answers is not in the document]
Comment (jengjeng) [reply api/forex/client.go:43]: yes, for the query builder.

File: api/forex/client.go (hunk 3 above)
Comment (jengjeng) [reply api/forex/client.go:160]: added type cast checking

File: api/forex/client.go (hunk 4 above)
Comment (jengjeng) [reply api/forex/client.go:165]: added type cast checking

File: websocket/websocket.go (hunk 6 above)
Comment (jengjeng) [reply websocket/websocket.go:19]: for managing the slice purpose (add, delete)

File: websocket/websocket.go (hunk 8 above)
Comment (jengjeng) [reply websocket/websocket.go:97]: removed os interrupt command. the loop will be gone.

File: api/forex/client.go (hunk 10 above)
Comment (jengjeng) [reply api/forex/client.go:156]: done

File: websocket/websocket.go (hunk 9 above)
Comment (jengjeng) [reply websocket/websocket.go:51]: updated. now you can pass full url likes `ws://localhost:4444/ws/forex`. but the query string will be placed anyway.

**Review by mnaichuk — APPROVED** [review]

**Pull request merged**